### DISCUSSION TITLE
compatible uncrustify version

### DISCUSSION THREAD

**Discussion by dhalbert** [micropython/micropython#14496] [Core Development] [answered]
Currently  uncrustify 0.71 or 0.72 is needed to match the settings in `tools/uncrustify.cfg`. @dlech kindly included a compatible version in the [pybricks PPA](https://launchpad.net/~pybricks/+archive/ubuntu/ppa). Later versions are not compatible

I recently upgraded to Ubuntu 24.04 to work out what changes in build procedure are needed to build CircuitPython. We run uncrustify as part of pre-commit. However, the standard uncrustify in the 24.04 packages is a lot newer and therefore incompatible, and the pybricks PPA does not (yet) have a build for 24.04.

I could build it myself, and even publish it somewhere for our users. But @dlech, are you planning to add 24.04 (Noble) to pybricks? We'd be grateful. Or does anyone else have a convenient package archive for this. Thanks.

Some related discussion here: https://github.com/micropython/micropython/pull/9980

**Comment by dhalbert** [comment]
I was able to download a .deb that intalled and ran from https://launchpad.net/~pybricks/+archive/ubuntu/ppa/+packages. Expand the Focal or Bionic download for uncrustify and there will be a list of files.

**Comment by dlech** [comment, marked as answer]
Instead of using a PPA going forward (would required manual configuration to force downgrade via apt pinning and only works with Ubuntu), I liked the suggestion of switching to a Python package. Would that work for you?

https://github.com/micropython/micropython/pull/9980#issuecomment-1317584925

**Reply by dhalbert** [reply]
That would be great, thanks. We have a `requirements-dev.txt` and would just add it to that.

**Reply by dlech** [reply]
Hi @dhalbert, I just published the package. So you can add `micropython-uncrustify` to your `requirements-dev.txt` now.

You can also check that it is working like this:

    pipx run --spec micropython-uncrustify uncrustify --version

**Reply by dhalbert** [reply]
@dlech Thank you!

**Reply by projectgus** [reply]
This is great, thanks @dlech!

I am curious, is there a reason why you started from 1.0.0 instead of 0.72 for the micropython-uncrustify version?

**Reply by dlech** [reply]
> a reason why you started from 1.0.0 instead of 0.72 for the micropython-uncrustify version?

In case we ever add a Python API to it. For example, we could move the MicroPython fixups there and run the whole thing as a standalone tool. And in my experience, there is always something unexpected that comes up when you try to use the version of one software project for the version of another software project, so I never try to do that anymore.

**Reply by projectgus** [reply]
Makes sense, thanks.
